projects/observability/src/shared/components/cartesian/d3/legend/cartesian-legend.ts

```diff
@@ -36,6 +40,13 @@ export class CartesianLegend {
     private readonly intervalData?: CartesianIntervalData,
     private readonly summaries: Summary[] = []
   ) {
+    this.numberOfgroups =
+      this.series.length === 0 || isNil(this.series[0].assignedGroup)
+        ? 0
+        : this.series[this.series.length - 1].assignedGroup!.id;
```

Comment:
`assignedGroup` is optional right? We shouldn't override it here.

Reply:
Actually, for this assigned group, if one of the series has this, every series will have this, This is the way group by working currently!! So It should be fine here

Reply:
We are introducing a tight coupling. If you keep it optional here then your app won't crash if the upstream stops sending `assignedGroup` obj (due to a bug).

Reply:
Also, why are you assigning `id` to number of groups?

Reply:
This reads to me like an assumption that the ids are numbers - we really shouldn't care, or even have ids for that matter. We also don't want to have magic indexing - easy source of bugs. We can just map through each series optional group name, collect them into a set. We have everything we need with just a group name on the series to organize, count distinct values etc.

Reply:
Removed `id` part from here!!